peps/pep-0759.rst

```diff
@@ -0,0 +1,472 @@
+PEP: 759
+Title: External Wheel Hosting
+Author: Barry Warsaw <barry@python.org>,
+        Ethan Smith <ethan@ethanhs.me>
+BDFL-Delegate: Donald Stufft <donald@python.org>
+Status: Draft
+Type: Standards Track
+Topic: Packaging
+Created: 19-Sep-2024
+Post-History:
+
+Abstract
+========
+
+This PEP proposes a mechanism by which projects hosted on `pypi.org
+<http://pypi.org>`__ can safely host wheel artifacts on external sites other
+than PyPI. This PEP explicitly does *not* propose external hosting of
+projects, packages, or their metadata. That functionality is already available
+by externally hosting independent package indexes. Because this PEP only
+provides a mechanism for projects to customize the download URL for specific
+released wheel artifacts, dependency resolution as already implemented by
+common installer tools such as `pip <https://pip.pypa.io/en/stable/>`__ and
+`uv <https://docs.astral.sh/uv/>`__ does not need to change.
+
+This PEP defines what it means to be "safe" in this context, along with a new
+package upload file format called a ``.rim`` file. It defines how ``.rim``
+files affect the metadata returned for a package's `Simple Repository API
+<https://packaging.python.org/en/latest/specifications/simple-repository-api/>`__
+in both HTML and JSON format, and how traditional wheels can easily be turned
+into ``.rim`` files.
+
+Rationale
+=========
+
+The Python Package Index, hosted at https://pypi.org, imposes `default limits
+<https://pypi.org/help/>`__ on upload artifact file size (100 MiB) and total
+project size (10.0 GiB). Most projects can comfortably fit within these limits
+during the lifetime of the project, through years of uploads. A few projects
+have encountered these limits, and have been granted both `file size
+exceptions <https://github.com/pypi/support/issues/4378>`__ and `project size
+exceptions <https://github.com/pypi/support/issues/4379>`__, allowing them to
+continue uploading new releases without having to take more drastic measures,
+such as removing files which may potentially still be in use by consumers
+(e.g. through version pins).
+
+A related workaround is the `"wheel stub"
+<https://github.com/wheel-next/wheel-stub>`__ approach, which provides an
+indirect link between PyPI and an external third party package index, where
+such limitations can be avoided. Wheel stubs are `source distributions
+<https://packaging.python.org/en/latest/specifications/source-distribution-format/>`__,
+a.k.a. "sdists" which utilize a :pep:`517` build backend that, instead of turning
+source code into a binary wheel, performs some logic to calculate the URL for
+an existing, externally hosted wheel to download.  Technically speaking, this
+satisfies the build backend contract, but the link between PyPI, the sdist,
+and the externally hosted wheel isn't evident, easily discoverable, or
+auditable.
```

Comment:
The reference to fulfilling the build backend contract seems a little out of context -- the point of this paragraph is that the behaviour is hidden and non-auditable, etc, rather than semantics of PEP 517. A suggestion to simplify:
```suggestion
an existing, externally hosted wheel to download and install.
This approach works, but the link between PyPI, the sdist,
and the externally hosted wheel isn't evident, easily discoverable, or
auditable.
```

Reply:
(context, I am the author of wheel-stub). I think the point here is that wheel-stub accomplishes the goals of this PEP (host a wheel externally), but in a sub-optimal manner. While it meets the letter of the standard for a build backend, it doesn't present information about where a wheel is coming from (since build backend info isn't visible by default in pip), or why. I'll chat with Barry and we will update this to be a bit clearer.

Reply:
Here's what I settled on:

```
A related workaround is the `"wheel stub" <https://github.com/wheel-next/wheel-stub>`__
approach, which provides an indirect link between PyPI and an external third party package
index, where such limitations can be avoided. Wheel stubs are :ref:`source distributions
<packaging:source-distribution-format>` (a.k.a. "sdists") which utilize a :pep:`517` build
backend that, instead of turning source code into a binary wheel, performs some logic to
calculate the URL for an existing, externally hosted wheel to download and install.  This
approach works, but it obscures the connection between PyPI, the sdist, and the externally
hosted wheel, since there is no way to present this information to ``pip`` or other such
tools.
```